LeanSAT/Model/Subst.lean

```diff
@@ -38,27 +46,137 @@ def PropForm.subst (p : PropForm ν₁) (f : ν₁ → PropForm ν₂) : PropFor
+  : vars (φ.subst f) = (vars φ).biUnion (fun v1 => vars (f v1)) := by
+  induction φ <;> simp [subst, Finset.biUnion_union, *]
 
 @[simp]
-theorem PropForm.satisfies_subst [DecidableEq ν] (φ : PropForm ν) (f : ν → PropForm ν') {τ : PropAssignment ν'}
-  : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ
-  := by induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+theorem satisfies_subst {φ : PropForm ν₁} {f} {τ : PropAssignment ν₂}
+    : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ := by
+  induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+
+-- CC: Better to use equivalence or ⟦ ⟧ = ⟦ ⟧?
+theorem subst_congr {φ₁ φ₂} (hφ : equivalent φ₁ φ₂)
+    : ∀ (σ : ν₁ → PropForm ν₂), (⟦φ₁.subst σ⟧ : PropFun _) = ⟦φ₂.subst σ⟧ := by
+  intro σ
+  apply PropFun.ext
+  intro τ
+  -- CC: simp_rw doesn't work here, for some reason...
+  rw [PropFun.satisfies_mk, PropFun.satisfies_mk, satisfies_subst, satisfies_subst,
+      ← PropFun.satisfies_mk, ← PropFun.satisfies_mk, Quotient.sound hφ]
+
+end subst /- section -/
+
+/-! ### `substOne` -/
+
+def substOne [DecidableEq ν] (φ : PropForm ν) (v : ν) (ψ : PropForm ν) : PropForm ν :=
+  φ.subst (fun v' => if v' = v then ψ else .var v')
+
+section substOne
+
+variable [DecidableEq ν] (φ φ₁ φ₂ : PropForm ν) (v : ν) (ψ ψ₁ ψ₂ : PropForm ν)
+
+theorem satisfies_substOne {φ : PropForm ν} {v} {τ : PropAssignment ν}
+    : τ ⊨ φ.substOne v ψ ↔ τ.set v (τ ⊨ ψ) ⊨ φ := by
+  simp [substOne]
+  apply iff_of_eq; congr
+  ext v
+  simp [PropAssignment.subst, PropAssignment.set]
+  split <;> simp
+  exact PropFun.satisfies_mk
+
+-- CC: Use `@Eq (PropFun _) ⟦ψ₁⟧ ⟦ψ₂⟧` or `equivalent φ₁ φ₂`?
+theorem substOne_congr {φ₁ φ₂ ψ₁ ψ₂} (v : ν) (hφ : @Eq (PropFun _) ⟦φ₁⟧ ⟦φ₂⟧) (hψ : @Eq (PropFun _) ⟦ψ₁⟧ ⟦ψ₂⟧)
+  : (⟦ φ₁.substOne v ψ₁ ⟧ : PropFun _) = ⟦ φ₂.substOne v ψ₂ ⟧ := by
+  apply PropFun.ext
+  intro τ
+  rw [PropFun.satisfies_mk, PropFun.satisfies_mk]
+  simp [substOne]
+  rw [← PropFun.satisfies_mk, ← PropFun.satisfies_mk, hφ]
+  apply iff_of_eq; congr; ext v
+  simp [PropAssignment.subst]
+  split
+  · rw [hψ]
+  · simp
+
+theorem vars_substOne : (PropForm.substOne φ v ψ).vars ⊆ (φ.vars \ {v}) ∪ ψ.vars := by
+  induction φ with
+  | var =>
+      intro v hv; simp [subst, substOne] at hv ⊢
+      split at hv
+      · simp [hv]
+      · simp at hv; subst_vars; simp [*]
+  | tr  => simp
+  | fls => simp
+  | neg φ₁ ih => simp; exact ih
+  | disj φ₁ φ₂ ih₁ ih₂
+  | conj φ₁ φ₂ ih₁ ih₂
+  | impl φ₁ φ₂ ih₁ ih₂
+  | biImpl φ₁ φ₂ ih₁ ih₂ =>
+    intro v hv; simp at *
+    cases hv
+    · have := ih₁ ‹_›
+      simp at this; rcases this with ⟨a,b⟩|c <;> simp [*]
+    · have := ih₂ ‹_›
+      simp at this; rcases this with ⟨a,b⟩|c <;> simp [*]
+
+end substOne /- section -/
+
+end PropForm
+
+/-! # PropFun -/
+
+namespace PropFun
+
+-- CC: A computable, lifting version of subst. Note the mis-ordering of arguments.
+def substL (f : ν₁ → PropForm ν₂) : PropFun ν₁ → PropFun ν₂ :=
```

Comment:
the diff between `subst` and `substL` (aside from computability) is the type of `f`. maybe add a doc comment explaining this?

Reply:
Left a doc comment, and matched the argument ordering to the other `subst` functions.